File: javascript/ch8/promise.js
const R = require('ramda');
const { IO } = require('../ch5/IO');
const average = require('./average');
const { getJSONPromise } = require('./getJSON');

function promise() {
  const log = R.curry((title, message) => console.log(title, message));

  getJSONPromise('http://localhost:5555/students')
  .then(R.map((student) => student.grade))
  .then(average)
  .then((grade) => IO.of(grade).map(log('promise success')).run())
  .catch((error) => log('promise fail', error));
}

module.exports = promise;

File: python/python/ch8/promise.py
import pydash

from ..ch5.IO import IO
from .average import average
from .get_json import get_json_promise

def promise():
    log = pydash.curry(lambda title, message: print(title, message))
    
    map = pydash.curry(lambda func, arr: pydash.map_(arr, func))
    
    (get_json_promise('http://localhost:5555/students')
     .then(map(lambda student: student['grade']))
     .then(average)
     .then(lambda grade: IO.of(grade).map(log('promise success')).run())
     .catch(lambda error: log('promise fail', error)))